gnome-initial-setup/demo/eos-demo-video.c

```diff
@@ -0,0 +1,230 @@
+
+  priv->main_window = GTK_WINDOW (gtk_application_window_new (GTK_APPLICATION (application)));
+  gtk_widget_set_events (GTK_WIDGET (priv->main_window),
+                         GDK_POINTER_MOTION_MASK |
+                         GDK_KEY_PRESS_MASK |
+                         GDK_BUTTON_PRESS_MASK); 
+  gtk_window_fullscreen (priv->main_window);
+  gtk_application_add_window (GTK_APPLICATION (application),
+                              priv->main_window);
+
+  /* Translators: this text appears when the user invokes
+   * some action that is blocked because the demo video
+   * is showing */
+  gtk_application_inhibit (GTK_APPLICATION (application),
+                           priv->main_window,
+                           GTK_APPLICATION_INHIBIT_LOGOUT |
+                           GTK_APPLICATION_INHIBIT_SWITCH |
+                           GTK_APPLICATION_INHIBIT_IDLE,
+                           _("Demonstration video"));
+
+  /* Reparent video into our application window */
+  video_parent = gtk_widget_get_parent (priv->widget);
+  if (video_parent)
+    {
+      gtk_container_remove (GTK_CONTAINER (video_parent), priv->widget);
+      gtk_widget_destroy (video_parent);
+    }
+
+  gtk_container_add (GTK_CONTAINER (priv->main_window), priv->widget);
+
+  gtk_widget_show (priv->widget);
+
+  g_signal_connect_object (priv->main_window,
+                           "event",
+                           G_CALLBACK (on_demo_video_event),
+                           demo_video,
+                           G_CONNECT_AFTER);
+}
+
+static void
+gis_demo_video_activate (GApplication *application)
+{
+  GisDemoVideo *demo_video = GIS_DEMO_VIDEO (application);
+  GisDemoVideoPrivate *priv = gis_demo_video_get_instance_private (demo_video);
+
+  G_APPLICATION_CLASS (gis_demo_video_parent_class)->activate (application);
+
+  g_timeout_add_seconds (2, on_show_application, demo_video);
+}
+
+static void
+gis_demo_video_init (GisDemoVideo *demo_video)
+{
+  GisDemoVideoPrivate *priv = gis_demo_video_get_instance_private (demo_video);
+  GstBus *bus = NULL;
+
+  priv->element = gst_element_factory_make ("gtksink", NULL);
+  g_object_get (&priv->element->object, "widget", &priv->widget, NULL);
+
+  priv->player = gst_element_factory_make ("playbin", "GisDemoVideoPlayBin");
+  priv->bus = gst_pipeline_get_bus (GST_PIPELINE (priv->player));
+
+  g_object_set (priv->player, "video-sink", g_object_ref_sink (priv->element), NULL);
+  g_object_set (priv->player, "uri", DEMO_VIDEO_FILE_URI, NULL);
+}
+
+static void
+gis_demo_video_dispose (GObject *object)
+{
+  GisDemoVideo *demo_video = GIS_DEMO_VIDEO (object);
+  GisDemoVideoPrivate *priv = gis_demo_video_get_instance_private (demo_video);
+
+
+  if (priv->player)
+    gst_element_set_state (priv->player, GST_STATE_NULL);
+
+  if (priv->bus)
+    {
+      gst_bus_set_flushing (priv->bus, TRUE);
+      gst_bus_remove_watch (priv->bus);
+      gst_object_replace ((GstObject **) &priv->bus, NULL);
```

Comment:
I think it should be possible to simply call `g_clear_object()` since `GstObject`s are also `GObject`s?